rfcs/20230801-max-partition-policy/README.md

```diff
@@ -0,0 +1,116 @@
+```cpp
+class graph {
+    ...
+    std::vector<partition> get_partitions(
+            partition::policy policy = partition::policy::fusion);
+    ...
+};
+```
+
+A partition policy is used in the API to control granularities
+of optimizations to be applied when partitioning.
+Currently, oneDNN Graph API supports two partition policies:
+
+```c
+typedef enum {
+    dnnl_graph_partition_policy_fusion = 1,
+    dnnl_graph_partition_policy_debug = 2,
+} dnnl_graph_partition_policy_t;
+```
+
+```cpp
+class partition {
+    ...
+    enum class policy {
+        fusion = dnnl_graph_partition_policy_fusion,
+        debug = dnnl_graph_partition_policy_debug,
+    };
+    ...
+};
+```
+
+- Fusion policy (default partition policy): fusion policy returns partitions
+  with [pre-defined fusion patterns](https://oneapi-src.github.io/oneDNN/dev_guide_graph_fusion_patterns.html),
+  e.g. Convolution followed by a chain of post-ops, MHA, MLP, etc.
+- Debug policy: debug policy returns partitions with single operation in each partition.
+  This policy doesn't apply any fusions, which is useful when
+  user wants to debug or compare performance with fusion policy.
+
+## Proposal
+
+The API to expose max policy to users is similar to fusion and debug policies.
+
+```c
+typedef enum {
+    dnnl_graph_partition_policy_max = 0, // the new max policy
+    dnnl_graph_partition_policy_fusion = 1,
+    dnnl_graph_partition_policy_debug = 2,
+} dnnl_graph_partition_policy_t;
+```
+
+```cpp
+class partition {
+    ...
+    enum class policy {
+        max = dnnl_graph_partition_policy_max, // the new max policy
+        fusion = dnnl_graph_partition_policy_fusion,
+        debug = dnnl_graph_partition_policy_debug,
+    };
+    ...
+};
+```
+
+The expected behaviors when using max policy:
+
+- If not specified, the default partition policy remains to be fusion policy.
+- Max policy returns partitions max in partition size. The criteria of partitioning
+  is based on pre-defined ops list and pre-defined rules. Connected ops
+  that meet the conditions of pre-defined ops list and pre-defined rules
+  will be selected into one partition.
+- The pre-defined ops list and pre-defined rules are backend specific.
+
+## Open questions
+
+1. Compared with fusion and debug policy, max policy gives the library more control
+   on partitioning. This on the other hand means users may lose fine-grained control
+   of what optimizations they want to get from the library.
+1. The performance of max policy may or may not be better than fusion policy. The
+   library tries to provide the best performance by model performance heuristics.
+   But for untested models, it is possible that the performance of max policy is
+   worse than fusion policy.
+
+## Known issues
+
+1. DNNL backend will not support max partition policy in oneDNN v3.3 release.
+   So when running with a combination of max partition policy and DNNL backend,
+   max partition policy will fallback to fusion partition policy.
```

Comment:
Though I'm not a big fan for implicit fallback, it seems to be the only solution for now to retain performance for max policy vs. fusion policy. Please clarify if this is short-term or a long-term solution even after DNNL backend supports the policy. And also clarify if this decision applies to all backends of the library.

Reply:
This is a long-term solution: for all backends that don't support max policy, it will fallback to fusion policy. And if a backend doesn't support fusion policy, it will fallback to debug policy.

I will add above statement in the rfc.